sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -51,12 +51,17 @@ trait DateTimeFormatterHelper {
   // In this way, synchronised is intentionally omitted in this method to make parallel calls
   // less synchronised.
   // The Cache.get method is not used here to avoid creation of additional instances of Callable.
-  protected def getOrCreateFormatter(pattern: String, locale: Locale): DateTimeFormatter = {
+  protected def getOrCreateFormatter(
+      pattern: String,
+      locale: Locale,
+      needVarLengthSecondFraction: Boolean = false): DateTimeFormatter = {
     val newPattern = DateTimeUtils.convertIncompatiblePattern(pattern)
-    val key = (newPattern, locale)
+    val useVarLen = needVarLengthSecondFraction && newPattern.split("'").zipWithIndex
+      .exists { case (p, idx) => idx % 2 == 0 && p.contains('S') }
```

Comment:
Also for this one, although it's only one line, maybe it still worth having a function in `DateTimeUtils` and more comments.

Reply:
`S` in literal is very rare, maybe `newPattern.contains('S')` is good enough as it's only for performance.

Reply:
aggreed